lib/developer_portal/lib/cms/toolbar.rb

```diff
@@ -39,6 +39,28 @@ def draft?
 
   private
 
+  def validate_and_extract_cms_token
+    signature = params.delete(:signature)
+    return signature if signature.blank?
+
+    expires_at = Time.at(params.delete(:expires_at).to_i).utc
+    raise ActiveSupport::MessageVerifier::InvalidSignature unless expires_at > Time.now.utc
+
+    verifier = Rails.application.message_verifier(:cms_token)
+    cms_token = site_account.settings.cms_token!
+    signing_options = { purpose: :cms_edit_mode, expires_at: }
+    valid_signature = signature == verifier.generate(cms_token.b, **signing_options).split("--").last
+
+    raise ActiveSupport::MessageVerifier::InvalidSignature unless valid_signature
+
+    cms_token
+  rescue ActiveSupport::MessageVerifier::InvalidSignature
+    # When the token is invalid, I don't think we have to bother the client and BugSnag with an exception
+    # Better return an empty token which means "Disable CMS edit mode (hide toolbar)"
+    flash[:error] = 'Disabling CMS edit mode due to an invalid or expired token'
```

Comment:
I think many would consider this an indication of a possible malicious activity. Perhaps we need to report it with low severity :thinking:

Reply:
I think just a log message that invalid token was tried for CMS portal X

Reply:
https://github.com/3scale/porta/pull/3938/commits/438ded035c0febc83b66e74fb5c170dbd0ac8c2e